lib/smf/smf.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright 2021 The Chromium OS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "smf.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(smf);
+
+/*
+ * Private structure (to this file) used to track state machine context.
+ * The structure is not used directly, but instead to cast the "internal"
+ * member of the smf_ctx structure.
+ */
+struct internal_ctx {
+	const struct smf_state *last_entered;
+	bool terminate : 1;
+	bool running   : 1;
+	bool enter     : 1;
+	bool exit      : 1;
+};
+
+#ifndef CONFIG_SMF_FLAT
+/* Gets the first shared parent state between a and b (inclusive) */
+static const struct smf_state *shared_parent_state(const struct smf_state *a,
+				const struct smf_state *b)
+{
+	const struct smf_state *orig_b = b;
+
+	/* There are no common ancestors */
+	if (b == NULL) {
+		return NULL;
+	}
+
+	/* This assumes that both A and B are NULL terminated without cycles */
+	while (a != NULL) {
+		/* We found a match return */
+		if (a == b) {
+			return a;
+		}
+
+		/*
+		 * Otherwise, increment b down the list for comparison until we
+		 * run out, then increment a and start over on b for comparison
+		 */
+		if (b->parent == NULL) {
+			a = a->parent;
+			b = orig_b;
+		} else {
+			b = b->parent;
+		}
+	}
+
+	return NULL;
+}
+#endif
+
+/*
+ * Call all entry functions of parents before children. If set_state is called
+ * during one of the entry functions, then do not call any remaining entry
+ * functions.
+ */
+static void call_entry_functions(struct smf_ctx *const ctx,
+					struct internal_ctx *const internal,
+			       const struct smf_state *stop,
+			       const struct smf_state *current)
+{
+	if (current == stop || internal->terminate) {
+		return;
+	}
+
+#ifndef CONFIG_SMF_FLAT
+	call_entry_functions(ctx, internal, stop, current->parent);
```

Comment:
Please no recursion.
The problem is that recursion can blow your stack easily.
It is also not obvious that for a flat state machine, that the recursion stops immediately because stop is NULL

Reply:
I've removed the recursion